Chore/upgrade prisma to version 6

timDeHof opened the pull request:
# Description

<!-- Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change. -->
This pull request includes several important updates and improvements across the codebase:

1. **Docker Configuration Update**

  - Updates base image to node:22-alpine (Updated from previous version [node:18-alpine])
  - Adds openssl installation for enhanced security and functionality

2. **Dependency Updates**

- Updates core NestJS packages (@nestjs/common, @nestjs/config, @nestjs/jwt) to version 11
![Screenshot 2025-03-21 at 9 19 33 PM](https://github.com/user-attachments/assets/103a16ab-e66b-4349-b7e9-596277523268)

- Updates Prisma client and related dependencies to latest version(6.5.0)  
![Screenshot 2025-03-21 at 9 18 47 PM](https://github.com/user-attachments/assets/2779497b-9f1b-488c-8181-e19d13d96ca7)

- Updates various other dependencies to their latest versions

3. **Ideation Management Improvements**

- Enhances manageOwnIdeationById function with improved role and permission checks
- Updates query structure for better performance and security
- Adds better error handling for team member ideation management
- Implements input validation through ValidationPipe in IdeationsModule
- Adds permission checks and data validation in IdeationsService

> [!IMPORTANT]
> you should run the command `docker compose down`, then `docker compose up -d --remove-orphans` to have an updated docker image    

# Useful Links:
[upgrading to prisma 6](https://www.prisma.io/docs/orm/more/upgrade-guides/upgrading-versions/upgrading-to-prisma-6)

[Announcing NestJS 11: What's New](https://trilon.io/blog/announcing-nestjs-11-whats-new)

## Issue link

Fixes # [CU-86b3zkvem](https://app.clickup.com/t/86b3zkvem)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Feature updates / changes
- [ ] Tests
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update


# How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

1. Docker Environment

- Verified successful build with new node:22-alpine base image by running the command: `node -v`
- Confirmed openssl installation and functionality
- Tested application deployment with new configuration

2. Dependency Updates

- Ran npm install to verify all dependencies install correctly, running `nest info`, `npx prisma --version`, and `npx prisma validate` 

- Executed test suite to ensure no breaking changes
- Verified application builds successfully with updated Nestjs packages

3. Ideation Management

- Tested permission checks for different user roles
- Verified validation pipe functionality
- Confirmed error handling for unauthorized access
- Tested data validation for ideation updates
- Validated updated query structure performance

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

timDeHof committed:
chore: update dependencies and Dockerfile for improved compatibility

- Updated Node.js version in Dockerfile to 20.1.0-alpine.
- Added OpenSSL installation in Dockerfile.
- Updated various dependencies in package.json, including @nestjs packages to version 11.x, @casl packages, and others for better performance and security.
- Adjusted yarn.lock to reflect the updated dependencies.
- Enhanced ideation management logic in ideations.ability.ts and ideations.service.ts for better permission handling.
- Integrated global validation pipe in ideations.module.ts for improved request validation.

timDeHof committed:
chore: update Dockerfile to use Node.js 22-alpine

- Changed base image in Dockerfile from node:20.1.0-alpine to node:22-alpine for improved compatibility.
- Moved yarn install command to follow OpenSSL installation for better build efficiency.

cherylli commented:
Task linked: [CU-86b3zkvem Upgrade Prisma to version 6](https://app.clickup.com/t/86b3zkvem)

cherylli commented:
I guess we also have to update the github actions?

![image](https://github.com/user-attachments/assets/46cdad38-3f68-4703-90d8-3512fad8bf4c)

cherylli commented:
Automatically closed by #252 (not intended)

timDeHof committed:
Merge branch 'dev' into chore/Upgrade-Prisma-to-version-6

cherylli commented on src/ideations/ideations.service.ts:
```diff
@@ -122,23 +122,25 @@ export class IdeationsService {
         const { title, description, vision } = updateIdeationDto;
 
-        await manageOwnIdeationById(req.user, ideationId);
+        // await manageOwnIdeationById(req.user, ideationId);
```
I think this can be removed

cherylli commented on src/ideations/ideations.service.ts:
```diff
@@ -122,23 +122,25 @@ export class IdeationsService {
+        // First check if user has permission to update this ideation
+        await manageOwnIdeationById(req.user, ideationId);
+
+        // Then validate the update data
```
I don't think this is validating the data, it's just checking if the data / dto exists. If the dto file is annotated, it should be automatically validated at the controller

timDeHof replied:
The changes were made because these ideations e2e tests were failing due to receiving a different error: 
```
● IdeationsController (e2e) › /PATCH /ideations/:ideationId › should return 403 when user tries to update someone else's ideation in the same team
    expected 403 "Forbidden", got 400 "Bad Request"

● IdeationsController (e2e) › /PATCH /ideations/:ideationId › should return 404 when ideation Id does not exist
    expected 404 "Not Found", got 400 "Bad Request"
```
The problem was that the error handling in the service layer is catching and transforming the errors before they reach the controller. 

So I moved `manageOwnIdeationById` check to be the first validation step and only check for update data validity after confirming the user has permission.

The previous implementation was failing because it was checking for valid update data before checking permissions, which meant that missing update data would always result in a 400 Bad Request error, even when the user didn't have permission to update the ideation or when the ideation didn't exist.

cherylli replied:
I see, do you know what change in the new nestjs version changes this? Since this previously would pass. I'm just curious

cherylli reviewed (commented):
Everything works, rebuilt my docker container ran all the tests etc

> Ideation Management Improvements
Enhances manageOwnIdeationById function with improved role and permission checks
Updates query structure for better performance and security
Adds better error handling for team member ideation management
Implements input validation through ValidationPipe in IdeationsModule
Adds permission checks and data validation in IdeationsService

Are these related to the version upgrade or just random changes?

> Implements input validation through ValidationPipe in IdeationsModule 

I see it being imported but where's the usage? I don't think it needs to be provided at the module level, since there are other custom validation pipes working in other modules without having to be provided.

timDeHof committed:
chore: removed unnecessary ideations module

- Removed global validation pipe from ideations module to streamline provider configuration.

timDeHof committed:
chore: update Dockerfile and README for OpenSSL troubleshooting

- Removed OpenSSL installation from Dockerfile to streamline the build process.
- Added troubleshooting section in README for Prisma not detecting OpenSSL, including instructions to run `yarn install` after spinning up Docker services.
